src/ssh/allowed_signer.rs

```
@@ -0,0 +1,289 @@
+            let (option_key, option_value) = match head.split_once('=') {
+                Some(v) => v,
+                None => (head, ""),
+            };
+            match option_key.to_lowercase().as_str() {
+                "cert-authority" => cert_authority = true,
+                "namespaces" => {
+                    if namespaces.is_some() {
+                        return Err(Error::InvalidAllowedSigner("multiple \"namespaces\" clauses".to_string()));
+                    }
+
+                    let (namespaces_value, rest_) = parse_namespaces(option_value, &mut rest)?;
+                    rest = rest_;
+                    let namespaces_value: Vec<&str> = namespaces_value.split(',')
+                        .filter(|e| !e.is_empty())
+                        .collect();
+                    namespaces = Some(
+                        namespaces_value.iter()
+                            .map(|s| s.to_string())
+                            .collect()
+                    );
+                },
+                "valid-after" => {
+                    if valid_after.is_some() {
+                        return Err(Error::InvalidAllowedSigner("multiple \"valid-after\" clauses".to_string()));
+                    }
+                    valid_after = Some(parse_timestamp(option_value)
+                        .map_err(|_| Error::InvalidAllowedSigner("invalid \"valid-after\" time".to_string()))?);
+                },
+                "valid-before" => {
+                    if valid_before.is_some() {
+                        return Err(Error::InvalidAllowedSigner("multiple \"valid-before\" clauses".to_string()));
+                    }
+                    valid_before = Some(parse_timestamp(option_value)
+                        .map_err(|_| Error::InvalidAllowedSigner("invalid \"valid-before\" time".to_string()))?);
+                },
+                // If option_key does not match any valid option, we test if it's the key data
```

I think it's better if you pull off the key data first then work on the options. That way if there are new option types in the future it doesn't have to break the parsing (we can choose to fail parsing with unknown options or allow)

---

Comments at the end of the line kinda complicates this... It's not easy to tell where the end starts without test parsing the key from the end to the front. I can't just slice off the last # either since that might be in the namespace.

---

Fair enough